---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 07-gene-set-analysis.md in _episodes_rmd/
source: Rmd
title: "Gene set analysis"
teaching: XX
exercises: XX
questions:
- "How do we find differentially expressed pathways?"
objectives:
- "Explain how to find differentially expressed pathways with gene set analysis in R."
- "Understand how differentially expressed genes can enrich a gene set."
- "Explain how to perform a gene set analysis in R, using clusterProfiler."
keypoints:
- "Key point 1"
---





Recall the differential expression analysis.


~~~
library(SummmarizedExperiment)
~~~
{: .language-r}



~~~
Error in library(SummmarizedExperiment): there is no package called 'SummmarizedExperiment'
~~~
{: .error}



~~~
library(DESeq2)
~~~
{: .language-r}


~~~
se <- readRDS("data/GSE96870_se.rds")
~~~
{: .language-r}


~~~
dds <- DESeq2::DESeqDataSet(se[, se$tissue == "Cerebellum"],
                            design = ~ sex + time)
~~~
{: .language-r}



~~~
Warning in DESeq2::DESeqDataSet(se[, se$tissue == "Cerebellum"], design = ~sex + : some variables in design formula are
characters, converting to factors
~~~
{: .warning}


~~~
dds <- DESeq2::DESeq(dds)
~~~
{: .language-r}

Fetch results for the contrast between male and female mice.


~~~
resSex <- DESeq2::results(dds, contrast = c("sex", "Male", "Female"))
summary(resSex)
~~~
{: .language-r}



~~~

out of 32652 with nonzero total read count
adjusted p-value < 0.1
LFC > 0 (up)       : 53, 0.16%
LFC < 0 (down)     : 71, 0.22%
outliers [1]       : 10, 0.031%
low counts [2]     : 13717, 42%
(mean count < 6)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results
~~~
{: .output}

Select differentially expressed (DE) genes between males and females
with FDR < 5%.


~~~
sexDE <- as.data.frame(subset(resSex, padj < 0.05))
dim(sexDE)
~~~
{: .language-r}



~~~
[1] 54  6
~~~
{: .output}



~~~
sexDE <- sexDE[order(abs(sexDE$log2FoldChange), decreasing=TRUE), ]
head(sexDE)
~~~
{: .language-r}



~~~
               baseMean log2FoldChange     lfcSE      stat        pvalue          padj
Eif2s3y       1410.8750       12.62514 0.5652155  22.33685 1.620659e-110 1.022366e-106
Kdm5d          692.1672       12.55386 0.5936267  21.14773  2.895664e-99  1.370011e-95
Uty            667.4375       12.01728 0.5935911  20.24505  3.927797e-91  1.486671e-87
Ddx3y         2072.9436       11.87241 0.3974927  29.86825 5.087220e-196 4.813782e-192
Xist         22603.0359      -11.60429 0.3362822 -34.50761 6.168523e-261 1.167393e-256
LOC105243748    52.9669        9.08325 0.5976242  15.19893  3.594320e-52  1.133708e-48
~~~
{: .output}



~~~
sexDEgenes <- rownames(sexDE)
head(sexDEgenes)
~~~
{: .language-r}



~~~
[1] "Eif2s3y"      "Kdm5d"        "Uty"          "Ddx3y"        "Xist"         "LOC105243748"
~~~
{: .output}



~~~
length(sexDEgenes)
~~~
{: .language-r}



~~~
[1] 54
~~~
{: .output}

Fetch chromosome location for all mouse genes and build a subset of them
located in the sex chrosomes X and Y.


~~~
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
allmousegenes <- genes(txdb)
sexGenesEntrez <- names(allmousegenes)[as.character(seqnames(allmousegenes)) %in% c("chrX", "chrY")]
head(sexGenesEntrez)
~~~
{: .language-r}



~~~
[1] "100034363" "100034729" "100038363" "100038584" "100038941" "100038977"
~~~
{: .output}



~~~
length(sexGenesEntrez)
~~~
{: .language-r}



~~~
[1] 1064
~~~
{: .output}

Further subset those mouse genes located in sex chromosomes to those for which we have
expression profiles.


~~~
sexGenesSE <- sexGenesEntrez[sexGenesEntrez %in% as.character(rowData(se)$ENTREZID)]
length(sexGenesSE)
~~~
{: .language-r}



~~~
[1] 1060
~~~
{: .output}

Because our subset of mouse genes in sex chromosomes are provided as Entrez identifiers,
let's build a table to match gene symbols to Entrez identifiers in our data.


~~~
sym2entrez <- rowData(se)$ENTREZID
names(sym2entrez) <- rownames(se)
~~~
{: .language-r}

Build a contingency table and conduct a one-tailed Fisher's exact test that verifies
the association between being DE and being located in a sex chromosome.


~~~
N <- nrow(se)
n <- length(sexDEgenes)
m <- length(sexGenesSE)
k <- length(intersect(sexGenesSE, sym2entrez[sexDEgenes])) 
dnames <- list(GS=c("inside", "outside"), DE=c("yes", "no"))
t <- matrix(c(k, n-k, m-k, N+k-n-m),
                        nrow=2, ncol=2, dimnames=dnames)
t
~~~
{: .language-r}



~~~
         DE
GS        yes    no
  inside   14  1046
  outside  40 40686
~~~
{: .output}



~~~
fisher.test(t, alternative="greater")
~~~
{: .language-r}



~~~

	Fisher's Exact Test for Count Data

data:  t
p-value = 5.303e-11
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 7.612423      Inf
sample estimates:
odds ratio 
  13.60958 
~~~
{: .output}
