---
source: Rmd
title: "Differential expression analysis"
teaching: XX
exercises: XX
questions:
- "How do we find differentially expressed genes?"
objectives:
- "Explain the steps involved in a differential expression analysis."
- "Explain how to perform these steps in R, using DESeq2."
keypoints:
- "Key point 1"
---

```{r, echo=FALSE, purl=FALSE, message=FALSE}
source("download_data.R")
```

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
options(width = 120)
```

> ## Contribute!
>
> This episode is intended to introduce the concepts required to perform 
> differential expression analysis with RNA-seq data. Explain concepts like 
> size factors, count modeling (Negative Binomial), dispersion, interpretation 
> of the test output, multiple testing correction. 
{: .callout}

```{r}
suppressPackageStartupMessages({
    library(SummarizedExperiment)
    library(DESeq2)
    library(ggplot2)
    library(ExploreModelMatrix)
    library(cowplot)
    library(ComplexHeatmap)
})
```

# Load data

```{r}
se <- readRDS("data/GSE96870_se.rds")
```

# Create DESeqDataSet 

```{r}
dds <- DESeq2::DESeqDataSet(se[, se$tissue == "Cerebellum"],
                            design = ~ sex + time)
```

# Run DESeq()

> ## Contribute!
>
> The concepts may be clearer if the steps of DESeq() are first performed 
> separately, followed by a note that they can be performed in a single step
> using DESeq().
{: .callout}

```{r}
dds <- DESeq2::DESeq(dds)
plotDispEsts(dds)
```

# Extract results for specific contrasts

> ## Contribute!
>
> Refer back to the episode about experimental design.
{: .callout}

```{r}
## Day 8 vs Day 0
resTime <- DESeq2::results(dds, contrast = c("time", "Day8", "Day0"))
summary(resTime)
head(resTime[order(resTime$pvalue), ])
DESeq2::plotMA(resTime)

## Male vs Female
resSex <- DESeq2::results(dds, contrast = c("sex", "Male", "Female"))
summary(resSex)
head(resSex[order(resSex$pvalue), ])
DESeq2::plotMA(resSex)
```

# Visualize selected set of genes

> ## Contribute!
>
> Here we intend to practice how to interpret the results from the 
> differential expression analysis. Refer back to the exploratory/QC episode.
{: .callout}

```{r}
vsd <- DESeq2::vst(dds, blind = TRUE)

genes <- rownames(head(resTime[order(resTime$pvalue), ], 10))
heatmapData <- assay(vsd)[genes, ]
heatmapData <- t(scale(t(heatmapData)))
heatmapColAnnot <- data.frame(colData(vsd)[, c("time", "sex")])
idx <- order(vsd$time)
heatmapData <- heatmapData[, idx]
heatmapColAnnot <- HeatmapAnnotation(df = heatmapColAnnot[idx, ])
ComplexHeatmap::Heatmap(heatmapData,
                        top_annotation = heatmapColAnnot,
                        cluster_rows = TRUE, cluster_columns = FALSE)
```

