---
source: Rmd
title: RStudio Project and Experimental Data
teaching: XX
exercises: XX
---

:::::::::::::::::::::::::::::::::::::: questions 

- How do you use RStudio project to manage your analysis project?
- What is the most effective way for organising directories for an analysis project?
- How to download a data set from Gene Expression Omnibus (GEO)?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Create an RStudio project and the necessary directories for carrying out an analysis.
- Download the data set that will be used for the subsequent episodes.

::::::::::::::::::::::::::::::::::::::::::::::::

## Managing an analysis project using RStudio projects

In this lesson, we will be using RStudio, a free integrated development environment (IDE),
to assist us in developing the R code necessary for analysing a RNAseq data.
Importantly, we will leverage RStudio projects, a feature built-in to RStudio for managing our analysis project.

:::::::::::::::::::::::::::::::::::::::  prereq

## Navigating RStudio
If you require assistance with RStudio or its general usage, please refer to our
[Introduction to data analysis with R and Bioconductor](https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html)
lesson.

:::::::::::::::::::::::::::::::::::::::

Typically, an analysis project starts with data set files, some R scripts, and 
output files in a directory. 
As the project evolves, this rapidly becomes more complex, with increasing 
scripts, output files, and their versions. 
If they aren't well-managed from the beginning, resuming the project after a 
break becomes challenging and time-consuming, as we struggle to recall the 
project's status and navigate the directory tree.
Without proper organisation, the project's complexity could lead to frequent 
use of the `setwd` function to switch between different working directories, 
resulting in a disorganized workspace.

:::::::::::::::::::::::::::::::::::::::  prereq

## What is a working directory?

A working directory in R is the default location on your computer where R looks 
for files to load or store any data you wish to save.
More information can be found in our 
[Introduction to data analysis with R and Bioconductor](https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html#the-working-directory)
lesson.

:::::::::::::::::::::::::::::::::::::::

### Structuring your working directory

For a more streamlined workflow, we suggest storing all files associated with 
an analysis in a specific directory, which will serve as your R working directory. 
Initially, this working directory should contain 4 distinct directories:

* `data`: dedicated to storing raw data. 
This folder should ideally only house raw data and not be modified unless you 
receive a new dataset (even then, if storage allows, we suggest you retain 
the previous dataset as well). 
For RNAseq data analysis, this directory will typically contain `*.fastq` files 
and any related metadata files for the experiment.
* `scripts`: for storing the R scripts you've written and utilised for analysing the data.
* `documents`: for storing documents related to your analysis, 
such as a manuscript outline or meeting notes with your team.
* `output`: for storing intermediate or final results generated by the R scripts
in the `scripts` directory. 
Importantly, if you carry out data cleaning or pre-processing, the output should 
ideally be stored in this directory, as these no longer represent raw data.

As your project grows in complexity, you might find it necessary to create more 
directories or sub-directories. 
Nevertheless, the aforementioned four directories should serve as the foundation 
of your working directory.

:::::::::::::::::::::::::::::::::::::::  challenge

## Create the directories for an RNAseq Analysis

Decide on the desired location for your working directory. 
Within this, create the 4 essential directories 
(`data`, `scripts`, `documents`, and `output`).

::::::::::::::::::::::::::::::::::: solution

The structure of your chosen working directory should resemble the following:

```{r, fig.cap="Your working directory should look like this", echo=FALSE, message=FALSE}
knitr::include_graphics("fig/02_setup/folders_working_dir.png")
```

:::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::

### Using RStudio project to manage your working directory

As stated earlier, RStudio project is a feature built-into RStudio for managing your analysis project.
It does so by storing project-specific settings within an `.Rproj` file located in your chosen working directory. 
Loading the settings up into RStudio by either opening the `.Rproj` file directly or through 
RStudio's open project option (from the menu bar, select `File` > `Open Project...`) will 
automatically set your R working directory to the 
location of the `.Rproj` file, essentially your chosen working directory.

Follow these steps to create a new RStudio project:

1. Start RStudio.
2. Navigate to the menu bar and select `File` > `New Project...`.
3. Choose `Existing Directory`.
4. Click `Browse...` button, and select the directory you have previously chosen
as the working directory for the analysis (i.e., the directory where the 4 essential
directories reside).
5. Click `Create Project` at the bottom right of the window.

Upon completion of the steps above, you'll find a `.Rproj` file within the directory 
previously selected as your working directory for the analysis. 

```{r, fig.cap="A new .Rproj file should be created in your chosen working directory.", echo=FALSE, message=FALSE}
knitr::include_graphics("fig/02_setup/rstudio_project.png")
```

Moreover, the heading of your RStudio console should now display the absolute path of 
the directory where `.Rproj` file resides, indicating that RStudio has set this directory as your 
R working directory.

```{r, fig.cap="Your R working directory should now be set to where the .Rproj file resides.", echo=FALSE, message=FALSE}
knitr::include_graphics("fig/02_setup/rstudio_console.png")
```

## Download data

The data we will use in this lesson is obtained from the [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo/), accession number [GSE96870](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96870).

```{r}
dir.create("data", showWarnings = FALSE)
download.file(
    url = "https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv", 
    destfile = "data/GSE96870_counts_cerebellum.csv"
)

download.file(
    url = "https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_cerebellum.csv?raw=true", 
    destfile = "data/GSE96870_coldata_cerebellum.csv"
)

download.file(
    url = "https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_all.csv?raw=true", 
    destfile = "data/GSE96870_coldata_all.csv"
)

download.file(
    url = "https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_rowranges.tsv?raw=true", 
    destfile = "data/GSE96870_rowranges.tsv"
)
```


::::::::::::::::::::::::::::::::::::: keypoints 

- When starting to use a new Bioconductor package, it is often good practice to go through the accompanying vignette and test some of the commands on the provided example data, before applying it to one's own data.

::::::::::::::::::::::::::::::::::::::::::::::::

