---
source: Rmd
title: Experimental design
teaching: XX
exercises: XX
editor_options:
  chunk_output_type: console
---

```{r, echo=FALSE, purl=FALSE, message=FALSE}
source("download_data.R")
```

::::::::::::::::::::::::::::::::::::::: objectives

- Explain the formula notation and design matrices.
- Explore different designs and learn how to interpret coefficients.

::::::::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::: questions

- How can one translate biological questions and comparisons to statistical terms suitable for use with RNA-seq analysis packages?

::::::::::::::::::::::::::::::::::::::::::::::::::

```{r, include=FALSE}
```

:::::::::::::::::::::::::::::::::::::::::  callout

### Contribute!

This episode is intended to provide a recap of the typical biological questions which would require an RNA-seq analysis approach; introduce what is an experimental design, then discuss the experimental designs in the context of the typical biological questions, interpretation of design parameters (treatmet, control, interaction, p-value, etc). 
It is important therfore to understand, what question to ask, which design to use, why that design, how to analysise the data and output.

::::::::::::::::::::::::::::::::::::::::::::::::::

```{r}
suppressPackageStartupMessages({
    library(SummarizedExperiment)
    library(ExploreModelMatrix)
    library(dplyr)
})
```

```{r}
meta <- read.csv("data/GSE96870_coldata_all.csv", row.names = 1)
meta
```

```{r}
vd <- VisualizeDesign(sampleData = meta, 
                      designFormula = ~ tissue + time + sex)
vd$cooccurrenceplots
```

## Compare males and females, non-infected spinal cord

```{r}
meta_noninf_spc <- meta %>% filter(time == "Day0" & 
                                       tissue == "Spinalcord")
meta_noninf_spc
vd <- VisualizeDesign(sampleData = meta_noninf_spc, 
                      designFormula = ~ sex)
vd$designmatrix
vd$plotlist
```

:::::::::::::::::::::::::::::::::::::::  challenge

### Challenge: Can you do it?

Set up the design formula to compare the three time points (Day0, Day4, Day8) in the male spinal cord samples, and visualize it using `ExploreModelMatrix`.

:::::::::::::::  solution

### Solution

```{r, echo=TRUE}
meta_male_spc <- meta %>% filter(sex == "Male" & tissue == "Spinalcord")
meta_male_spc
vd <- VisualizeDesign(sampleData = meta_male_spc, designFormula = ~ time)
vd$designmatrix
vd$plotlist
```

:::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::

## Factorial design without interactions

```{r}
meta_noninf <- meta %>% filter(time == "Day0")
meta_noninf
vd <- VisualizeDesign(sampleData = meta_noninf, 
                      designFormula = ~ sex + tissue)
vd$designmatrix
vd$plotlist
```

## Factorial design with interactions

```{r}
meta_noninf <- meta %>% filter(time == "Day0")
meta_noninf
vd <- VisualizeDesign(sampleData = meta_noninf, 
                      designFormula = ~ sex * tissue)
vd$designmatrix
vd$plotlist
```

## Paired design

```{r}
meta_fem_day0 <- meta %>% filter(sex == "Female" & 
                                     time == "Day0")
meta_fem_day0
vd <- VisualizeDesign(sampleData = meta_fem_day0,
                      designFormula = ~ mouse + tissue)
vd$designmatrix
vd$plotlist
```

## Within- and between-subject comparisons

```{r}
meta_fem_day04 <- meta %>% 
    filter(sex == "Female" & 
               time %in% c("Day0", "Day4")) %>%
    droplevels()
meta_fem_day04
design <- model.matrix(~ mouse, data = meta_fem_day04)
design <- cbind(design, 
                Spc.Day0 = meta_fem_day04$tissue == "Spinalcord" & 
                    meta_fem_day04$time == "Day0",
                Spc.Day4 = meta_fem_day04$tissue == "Spinalcord" & 
                    meta_fem_day04$time == "Day4")
rownames(design) <- rownames(meta_fem_day04)
design
vd <- VisualizeDesign(sampleData = meta_fem_day04 %>%
                          select(time, tissue, mouse),
                      designFormula = NULL, 
                      designMatrix = design, flipCoordFitted = FALSE)
vd$designmatrix
vd$plotlist
```

:::::::::::::::::::::::::::::::::::::::: keypoints

- The formula framework in R allows creation of design matrices, which details the variables expected to be associated with systematic differences in gene expression levels. 
- Comparisons of interest can be defined using contrasts, which are linear combinations of the model coefficients. 


::::::::::::::::::::::::::::::::::::::::::::::::::


